hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/grpc/GrpcOmServerRequestInterceptor.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.om.grpc;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.hadoop.ozone.om.grpc.metrics.GrpcOzoneManagerMetrics;
+import io.grpc.ForwardingServerCallListener.SimpleForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectOutputStream;
+import java.io.InputStream;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * Interceptor to gather metrics based on grpc server request.
+ */
+public class GrpcOmServerRequestInterceptor implements ServerInterceptor {
+
+  private final GrpcOzoneManagerMetrics grpcMetrics;
+  private long bytesReceived;
+  private long receivedTime;
+  private long startTime;
+  private long endTime;
+
+  public GrpcOmServerRequestInterceptor(
+      GrpcOzoneManagerMetrics grpcMetrics) {
+    super();
+    this.grpcMetrics = grpcMetrics;
+    this.bytesReceived = 0;
+  }
+
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> serverCall, Metadata headers,
+      ServerCallHandler<ReqT, RespT> serverCallHandler) {
+
+    // received time
+    receivedTime = System.nanoTime();
+
+    return new SimpleForwardingServerCallListener<ReqT>(
+        serverCallHandler.startCall(serverCall, headers)) {
+
+      @Override
+      public void onMessage(ReqT message) {
+        // start time
+        startTime = System.nanoTime();
+
+        ByteArrayOutputStream byteArrayOutputStream =
+            new ByteArrayOutputStream();
+        try {
+          ObjectOutputStream outputStream =
+              new ObjectOutputStream(byteArrayOutputStream);
+          outputStream.writeObject(message);
```

Comment:
@duongkame I can't find a better way to do this and gathering these metrics on the client-side, doesn't seem the right approach. It also makes it difficult to reuse the metrics code for SCM or DN. My understanding is that interceptors propagate the call, don't block it. The system doesn't wait for this calculation to complete to move on to process the next message.

Reply:
Ok, so metrics collection is not blocking. But we're doubling the cost of deserialization/serialization just to collect `bytesReceived`/`bytesSent` metrics. This still doesn't a lot of sense to me.
If there's technical collecting messages' size, we'd rather have counters like `messagesReceived`/`messagesSent` instead. I think they're comparably useful and don't indicate additional cost.

Reply:
Thanks, @xBis7, and sorry that I lost focus on this PR.

I see you've changed to use `message.toString()` to count the bytes.

Actually, I did a bit of homework to trace the actual type of such message objects passed to the GRPC listeners. It looks to me that when using GRPC with Protobuf, the messages are parsed to Protobuf-generated classes, e.g. `AllocateBlockRequest` or `AllocateBlockResponse`. And those classes extend the same class `com.google.protobuf.AbstractMessage`, which is a lot easier to extract information like message size.

So, if we check and safely cast to `AbstractMessage`, the message size can be extracted like below:
```
int getMessageSize(Object message) {
   if (message instanceof AbstractMessage) {
      AbstractMessage parsedMessage = (AbstractMessage) message;
      return parsedMessage.getSerializedSize();
   }
  // otherwise maybe just return 0, or fall back calculation.
}
```

Reply:
@duongkame Thanks for looking into this. As a fall back, I will use the first approach where we are serializing the message because the results are the same with `parsedMessage.getSerializedSize()`.

Reply:
I think maybe we can just skip the operation and emit a metric to indicate Ozone cannot capture the details.

Reply:
@duongkame I fixed it!